restrict_download_region/restrict_region.py

```diff
@@ -0,0 +1,153 @@
+    resource_properties = event.get('ResourceProperties', {})
+    bucket_name = resource_properties.get('BucketName')
+    if not bucket_name:
+        raise ValueError("Bucket name is missing")
+    return bucket_name
+
+
+def generate_ip_address_policy(bucket_name: str, region_ip_prefixes: List[str]):
+    return {
+        'Sid': POLICY_STATEMENT_ID,
+        'Effect': 'Deny',
+        'Principal': '*',
+        'Action': 's3:GetObject',
+        'Resource': 'arn:aws:s3:::'+bucket_name+'/*',
+        'Condition': {
+            'NotIpAddress': {'aws:SourceIp': region_ip_prefixes},
+            # allows any S3 VPC Endpoint to bypass the ip restriction.
+            # cross region gateway endpoints are not supported in AWS so any S3 VPC endpoint
+            # traffic is implicitly same region.
+            'Null': {'aws:sourceVpc': 'true'}
+        }
+    }
+
+
+def get_ip_prefixes_for_region() -> List[str]:
+    if not REGION:
+        raise ValueError("REGION must be defined.")
+
+    # generate new policy statement based on data from AWS
+    all_ip_prefixes = requests.get('https://ip-ranges.amazonaws.com/ip-ranges.json').json()
+    return ip_prefixes_for_region(all_ip_prefixes['prefixes'], 'ip_prefix', REGION) + \
+        ip_prefixes_for_region(all_ip_prefixes['ipv6_prefixes'], 'ipv6_prefix', REGION)
+
+
+def get_bucket_policy(s3_client, bucket_name: str) -> dict:
+    try:
+        return json.loads(s3_client.get_bucket_policy(Bucket=bucket_name)['Policy'])
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchBucketPolicy':
+            return {
+                "Version": "2012-10-17",
+                "Statement": []
+            }
+        else:
+            raise
+
+
+def process_ip_restrict_policy(bucket_name: str, region_ip_prefixes: List[str], custom_resource_request_type: str, bucket_policy: dict):
+    """
+    Modifies the passed in bucket_policy and decides whether to add or remove the IP restriciting policy
+    """
+    # filter out the previously set IP filtering policy
+    bucket_policy['Statement'] = [statement for statement in bucket_policy['Statement']
+                                  if (POLICY_STATEMENT_ID != statement.get("Sid"))]
+
+    # skip adding new policy if deleting the custom resource
+    if custom_resource_request_type == 'Delete':
+        return
+
+    # add new IP address policy statement
+    new_ip_policy_statement = generate_ip_address_policy(bucket_name, region_ip_prefixes)
+    bucket_policy['Statement'].append(new_ip_policy_statement)
+
+
+def update_bucket_policy(s3_client, bucket_name: str, bucket_policy: dict):
+    # update existing policy or delete policy completely if there are no longer any policies left
+    if bucket_policy['Statement']:
+        s3_client.put_bucket_policy(Bucket=bucket_name, Policy=json.dumps(bucket_policy))
+    else:
+        s3_client.delete_bucket_policy(Bucket=bucket_name)
```

Comment:
`process_ip_restrict_policy()` will preserve all other unrelated policies, so the only possible time delete is used is if we had no other policies: 
```
{
    "Version": "2012-10-17",
    "Statement": []
}
```
S3 used to allow us to `PUT` this policy, but now it will throw error stating the policy is malformed.